clients/src/main/java/org/apache/kafka/clients/admin/NewPartitionReassignment.java

```diff
@@ -21,14 +21,24 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 
 /**
  * A new partition reassignment, which can be applied via {@link AdminClient#alterPartitionReassignments(Map, AlterPartitionReassignmentsOptions)}.
  */
 public class NewPartitionReassignment {
     private final List<Integer> targetReplicas;
 
-    public NewPartitionReassignment(List<Integer> targetReplicas) {
+    /**
+     * @throws IllegalArgumentException if no replicas are supplied
+     */
+    public static Optional<NewPartitionReassignment> of(List<Integer> replicas) {
```

Comment:
This is a pretty strange method. Why would you have a public method that always return a non empty `Optional`?

Reply:
We discussed it at length here (https://github.com/apache/kafka/pull/7570#discussion_r337384275) and offline. This is a nifty helper method that reduces boilerplate for users when composing the AdminClient API call, since the `AlterPartitionReassignments` accepts a map of `TopicPartition, Optional<NewPartitionReassignment>`. 

Because we do not expect any other uses of `NewPartitionReassignment` outside of an optional,  I thought it was reasonable to return this.

Reply:
I don't see much of a discussion in that comment. :) Generally, `Optional` is meant to be returned when a method can return a value or not return a value. Here, we're always wrapping the instance. Furthermore, `of` is typically used as a static factory method. We also made the constructor private, so one cannot create an instance without going through this non standard and somewhat confusing method.

Reply:
With relation to the private constructor, the idea was that you would never need to construct a `NewPartitionReassignment` that is not inside an Optional. I thought it made sense to nudge people to use the helper.

Perhaps we are breaking convention with the naming (`of`) and the return of an `Optional`. Is there a way to both reduce boilerplate and keep conventions intact in this public API, or do we prefer to trade off verbosity for convention?

I personally wouldn't have been confused and just see this as a helper, but I don't have an extensive Java background.

Reply:
If most users will in fact experience more confusion than benefit, it makes sense we remove it. Are we too late for that?

Reply:
Do we have evidence that this actually reduces boilerplate in a meaningful way in real code? I personally doubt it. In tests, there's a slight benefit. When it comes to public APIs, when in doubt, keep it out. You can always add it later. It's not too late since RC0 hasn't been created yet. cc @cmccabe @omkreddy

Reply:
Depends on how we define meaningful. Given that it's saving a constructor call and an Optional wrap, it may not be that much.

Here is an example of what client code would look like without:
```
        Map<TopicPartition, List<Integer>> proposedReassignments = new HashMap<>();
        Map<TopicPartition, Optional<NewPartitionReassignment>> partitionsToReassign = new HashMap<>();
        for (Map.Entry<TopicPartition, List<Integer>> proposedReassignment : proposedReassignments.entrySet()) {
            partitionsToReassign.put(
                proposedReassignment.getKey(),
                Optional.of(new NewPartitionReassignment(proposedReassignment.getValue()))
            );
        }
```

And here it is with it:
```
Map<TopicPartition, List<Integer>> proposedReassignments = new HashMap<>();
        Map<TopicPartition, Optional<NewPartitionReassignment>> partitionsToReassign = new HashMap<>();
        for (Map.Entry<TopicPartition, List<Integer>> proposedReassignment : proposedReassignments.entrySet()) {
            partitionsToReassign.put(
                proposedReassignment.getKey(),
                NewPartitionReassignment.of(proposedReassignment.getValue())
            );
        }
```

You could say that `new NewPartitionReassignment` reads weird as well. Perhaps we could have done better with the naming of the class

Reply:
In the former, you can handle the Optional.empty() where the latter does not? I'd even say the latter code is less clear from that perspective.

Reply:
As in, the latter code does not make it explicit that we are passing in an optional?
The value can be replaced with `Optional.empty()` in both situations. I'm not sure I understand

Reply:
None of us felt very strongly about this method and I defaulted to the one that seemed more pleasant to me.
Since we are in doubt of how useful and confusing it'll be, it makes sense to keep it out for this next release. I've opened a PR to get this right - https://github.com/apache/kafka/pull/7638

Reply:
I just mean the code snippet you passed doesn't handle removals, which is the whole reason for passing `Optional`.

Reply:
For what its worth, I had to use this api in ADB changes and I didn't like it. It made me go through hoops to get things done. Here is PR if someone is interested to see usage in real world use case: https://github.com/confluentinc/confluent-rebalancer/pull/21

I would prefer a traditional api (either make C'tor private and `of` as a builder pattern or provide C'tors to create objects, but don't return Optional).